    This dataset contains inventory related information such as Backlog, Safety Stock, Projected Stock and so on for over 13 months. Our stakeholders wanted to check if any of their materials are in danger of going into Out of Stock (OOS).
    
    
We consider a material OOS if the Projected Stock goes is below 0 during any of the weeks listed in the files. One important information for procurement is to learn when is the first week the given material goes to OOS. It is also beneficial to know how many times does this happen if no intervention happens    .    

Ta    sk:
-	Load all the data and filter for Projected Stock and for rows where any of the week (W29/2021, W30/2021 … Wn/2021) goes bel    ow 0
-	Indicate which is the first week that this ha    ppens
-	Count the number of OOS weeks for each Material


In [82]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings 
warnings.filterwarnings("ignore")

In [83]:
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.0f}'.format)
pd.set_option('display.max_seq_items', None)

In [84]:
plant_a = pd.read_csv(r"C:\Users\lily\Desktop\DS interview prep\Givudan\Exercise3\Resources\Plant A.csv")
plant_b = pd.read_csv(r"C:\Users\lily\Desktop\DS interview prep\Givudan\Exercise3\Resources\Plant B.csv")
plant_c = pd.read_csv(r"C:\Users\lily\Desktop\DS interview prep\Givudan\Exercise3\Resources\Plant C.csv")

In [85]:
plant_a.head()

Material Plant               Values  Stock  Backlog  W29/2021  W30/2021  \
0 538225000000  PMCX        Initial Stock  23266        0         0         0   
1 538225000000  PMCX         Safety Stock      0        0         0         0   
2 538225000000  PMCX   Total Requirements      0    -2253     -8301     -4379   
3 538225000000  PMCX  Total Replenishment      0        0         0      6000   
4 538225000000  PMCX           In Transit      0        0         0         0   

   W31/2021  W32/2021  W33/2021  W34/2021  W35/2021  W36/2021  W37/2021  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2     -6863     -7035     -7529     -5443     -3499     -1219     -2004   
3      6000         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   W38/2021  W39/2021  W40/2021  W41/2021  
0         0         0         0         0  
1         0         0         0         0  
2     -3996     -2343     -2617      -660  
3         0         0         0         0  
4         0         0         0         0

In [86]:
plant_a['Values'].value_counts()

Values
Initial Stock                  273
Safety Stock                   273
Total Requirements             273
Total Replenishment            273
In Transit                     273
Projected Stock                273
Projected Stock w/o backlog    273
3rd party vendor               273
Affiliate                      273
Name: count, dtype: int64

In [87]:
plant_a['Plant'].value_counts()

Plant
PMCX    2457
Name: count, dtype: int64

In [88]:
plant_b.head()

Material Plant               Values  Stock  Backlog  W29/2021  W30/2021  \
0 627154000000  4QCF        Initial Stock    327        0         0         0   
1 627154000000  4QCF         Safety Stock      0        0         0         0   
2 627154000000  4QCF   Total Requirements      0        0       -52        -5   
3 627154000000  4QCF  Total Replenishment      0        0         0         0   
4 627154000000  4QCF           In Transit      0        0         0         0   

   W31/2021  W32/2021  W33/2021  W34/2021  W35/2021  W36/2021  W37/2021  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2       -99         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   W38/2021  W39/2021  W40/2021  W41/2021  
0         0         0         0         0  
1         0         0         0         0  
2         0        -0        -1        -1  
3         0         0         0         0  
4         0         0         0         0

In [89]:
plant_b['Values'].value_counts()

Values
Initial Stock                  1500
Safety Stock                   1500
Total Requirements             1500
Total Replenishment            1500
In Transit                     1500
Projected Stock                1500
Projected Stock w/o backlog    1500
3rd party vendor               1500
Affiliate                      1500
Name: count, dtype: int64

In [90]:
plant_b['Plant'].value_counts()

Plant
4QCF    13500
Name: count, dtype: int64

In [91]:
plant_c.head() 

Material Plant               Values  Stock  Backlog  W29/2021  W30/2021  \
0 255958000000  4QCF        Initial Stock     28        0         0         0   
1 255958000000  4QCF         Safety Stock      0        0         0         0   
2 255958000000  4QCF   Total Requirements      0       -0        -2        -2   
3 255958000000  4QCF  Total Replenishment      0        0         0         0   
4 255958000000  4QCF           In Transit      0        0         0         0   

   W31/2021  W32/2021  W33/2021  W34/2021  W35/2021  W36/2021  W37/2021  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2        -1        -1        -1        -1        -1        -1        -1   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   W38/2021  W39/2021  W40/2021  W41/2021  
0         0         0         0         0  
1         0         0         0         0  
2        -1        -1        -1        -1  
3         0         0         0         0  
4         0         0         0         0

In [92]:
plant_c['Values'].value_counts()

Values
Initial Stock                  1369
Safety Stock                   1369
Total Requirements             1369
Total Replenishment            1369
In Transit                     1369
Projected Stock                1369
Projected Stock w/o backlog    1369
3rd party vendor               1369
Affiliate                      1369
Name: count, dtype: int64

In [93]:
plant_c['Plant'].value_counts()

Plant
4QCF    12321
Name: count, dtype: int64

*Note: Same plant code for Plant B and Plant C?*

In [94]:
dataframes = [plant_a, plant_b, plant_c]
table_names = ['Plant A.csv', 'Plant B.csv', 'Plant C.csv']

In [95]:
for df, name in zip(dataframes, table_names):
    df['Source'] = name

In [96]:
df = pd.concat(dataframes, ignore_index=True)
df.head()

Material Plant               Values  Stock  Backlog  W29/2021  W30/2021  \
0 538225000000  PMCX        Initial Stock  23266        0         0         0   
1 538225000000  PMCX         Safety Stock      0        0         0         0   
2 538225000000  PMCX   Total Requirements      0    -2253     -8301     -4379   
3 538225000000  PMCX  Total Replenishment      0        0         0      6000   
4 538225000000  PMCX           In Transit      0        0         0         0   

   W31/2021  W32/2021  W33/2021  W34/2021  W35/2021  W36/2021  W37/2021  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2     -6863     -7035     -7529     -5443     -3499     -1219     -2004   
3      6000         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   W38/2021  W39/2021  W40/2021  W41/2021       Source  
0         0         0         0         0  Plant A.csv  
1         0         0         0         0  Plant A.csv  
2     -3996     -2343     -2617      -660  Plant A.csv  
3         0         0         0         0  Plant A.csv  
4         0         0         0         0  Plant A.csv

In [97]:
filtered_ps = df[df['Values'] == 'Projected Stock']
filtered_ps.head()

Material Plant           Values  Stock  Backlog  W29/2021  W30/2021  \
5  538225000000  PMCX  Projected Stock  23266    21013     12712     14333   
14 277881000000  PMCX  Projected Stock  12725    12725     10025      5525   
23 264543000000  PMCX  Projected Stock   3710     3464      2323      1531   
32 858173000000  PMCX  Projected Stock   3475     2958      4334      3956   
41 250362000000  PMCX  Projected Stock    339      339       126       126   

    W31/2021  W32/2021  W33/2021  W34/2021  W35/2021  W36/2021  W37/2021  \
5      13470      6435     -1093     -6537    -10036    -11255    -13259   
14      5525      5525      5525      5525      2825      2825      2825   
23      1257      2244      1253       857       857       632       632   
32      3644      2357      2167      2080      1918      3406      2131   
41        91        91        91        56        56        56        56   

    W38/2021  W39/2021  W40/2021  W41/2021       Source  
5     -17255    -19598    -22214    -22874  Plant A.csv  
14      2825      -775      -775      -775  Plant A.csv  
23       632       632      -161      -161  Plant A.csv  
32      1969      1882      1870      1645  Plant A.csv  
41        56        56        56        56  Plant A.csv

In [98]:
len(filtered_ps)

3142

In [99]:
week_col = []

for col in df.columns:
    if col.startswith('W'):
        week_col.append(col)

In [100]:
week_col

['W29/2021',
 'W30/2021',
 'W31/2021',
 'W32/2021',
 'W33/2021',
 'W34/2021',
 'W35/2021',
 'W36/2021',
 'W37/2021',
 'W38/2021',
 'W39/2021',
 'W40/2021',
 'W41/2021']

In [101]:
filtered_df = filtered_ps[(filtered_ps[week_col] < 0).any(axis=1)]

In [102]:
filtered_df.head()

Material Plant           Values  Stock  Backlog  W29/2021  W30/2021  \
5   538225000000  PMCX  Projected Stock  23266    21013     12712     14333   
14  277881000000  PMCX  Projected Stock  12725    12725     10025      5525   
23  264543000000  PMCX  Projected Stock   3710     3464      2323      1531   
59  447759000000  PMCX  Projected Stock   9596     9596      9596      6876   
113 932197000000  PMCX  Projected Stock  40131    36751     36751     26426   

     W31/2021  W32/2021  W33/2021  W34/2021  W35/2021  W36/2021  W37/2021  \
5       13470      6435     -1093     -6537    -10036    -11255    -13259   
14       5525      5525      5525      5525      2825      2825      2825   
23       1257      2244      1253       857       857       632       632   
59      -3784      9840      9840      9840      9840      9840      9840   
113     20526     12236     20916     17966     12066     12066      3776   

     W38/2021  W39/2021  W40/2021  W41/2021       Source  
5      -17255    -19598    -22214    -22874  Plant A.csv  
14       2825      -775      -775      -775  Plant A.csv  
23        632       632      -161      -161  Plant A.csv  
59       9840      9840      7120      7120  Plant A.csv  
113     -2124     -5074     -5134    -11034  Plant A.csv

In [103]:
len(filtered_df)

278

*Note: the funal number of rows are 278 - these materials has below zero value for projected stock - during any of the weeks.*

In [104]:
def first_oos(material):
    for week in week_col:
        if material[week] < 0:
            return week
    return None

In [105]:
filtered_df = filtered_df.copy() 

In [106]:
filtered_df.loc[:, 'OOS week'] = filtered_df.apply(first_oos, axis=1)

In [107]:
filtered_df.head()

Material Plant           Values  Stock  Backlog  W29/2021  W30/2021  \
5   538225000000  PMCX  Projected Stock  23266    21013     12712     14333   
14  277881000000  PMCX  Projected Stock  12725    12725     10025      5525   
23  264543000000  PMCX  Projected Stock   3710     3464      2323      1531   
59  447759000000  PMCX  Projected Stock   9596     9596      9596      6876   
113 932197000000  PMCX  Projected Stock  40131    36751     36751     26426   

     W31/2021  W32/2021  W33/2021  W34/2021  W35/2021  W36/2021  W37/2021  \
5       13470      6435     -1093     -6537    -10036    -11255    -13259   
14       5525      5525      5525      5525      2825      2825      2825   
23       1257      2244      1253       857       857       632       632   
59      -3784      9840      9840      9840      9840      9840      9840   
113     20526     12236     20916     17966     12066     12066      3776   

     W38/2021  W39/2021  W40/2021  W41/2021       Source  OOS week  
5      -17255    -19598    -22214    -22874  Plant A.csv  W33/2021  
14       2825      -775      -775      -775  Plant A.csv  W39/2021  
23        632       632      -161      -161  Plant A.csv  W40/2021  
59       9840      9840      7120      7120  Plant A.csv  W31/2021  
113     -2124     -5074     -5134    -11034  Plant A.csv  W38/2021

In [108]:
filtered_df = filtered_df.copy() 

In [109]:
filtered_df.loc[:, 'No of OOS weeks'] = (filtered_df[week_col] < 0).sum(axis=1)

In [110]:
filtered_df

Material Plant           Values  Stock  Backlog  W29/2021  W30/2021  \
5     538225000000  PMCX  Projected Stock  23266    21013     12712     14333   
14    277881000000  PMCX  Projected Stock  12725    12725     10025      5525   
23    264543000000  PMCX  Projected Stock   3710     3464      2323      1531   
59    447759000000  PMCX  Projected Stock   9596     9596      9596      6876   
113   932197000000  PMCX  Projected Stock  40131    36751     36751     26426   
...            ...   ...              ...    ...      ...       ...       ...   
28049 140134000000  4QCF  Projected Stock   9720     9720      4860      4860   
28067 273497000000  4QCF  Projected Stock  11987     3247      3247     -2843   
28076 115969000000  4QCF  Projected Stock      0        0         0         0   
28211 388052000000  4QCF  Projected Stock   1025      385       385       385   
28265   1259168724  4QCF  Projected Stock      0       -6        -6        -6   

       W31/2021  W32/2021  W33/2021  W34/2021  W35/2021  W36/2021  W37/2021  \
5         13470      6435     -1093     -6537    -10036    -11255    -13259   
14         5525      5525      5525      5525      2825      2825      2825   
23         1257      2244      1253       857       857       632       632   
59        -3784      9840      9840      9840      9840      9840      9840   
113       20526     12236     20916     17966     12066     12066      3776   
...         ...       ...       ...       ...       ...       ...       ...   
28049      4860     13500     13500      4860      4860      4860      4860   
28067     -2843     -2843     -6443     -6443     -6443     -6443     -7133   
28076         0         0         0         0         0         0         0   
28211       385       385       385       385       385      -255       145   
28265        -6        -6       -12       -12       -12       -12       -18   

       W38/2021  W39/2021  W40/2021  W41/2021       Source  OOS week  \
5        -17255    -19598    -22214    -22874  Plant A.csv  W33/2021   
14         2825      -775      -775      -775  Plant A.csv  W39/2021   
23          632       632      -161      -161  Plant A.csv  W40/2021   
59         9840      9840      7120      7120  Plant A.csv  W31/2021   
113       -2124     -5074     -5134    -11034  Plant A.csv  W38/2021   
...         ...       ...       ...       ...          ...       ...   
28049      4860         0     -8640     -8640  Plant C.csv  W40/2021   
28067    -13223    -13223    -17603    -18293  Plant C.csv  W30/2021   
28076         0         0         0      -915  Plant C.csv  W41/2021   
28211       145       145       145       145  Plant C.csv  W36/2021   
28265       -26       -26       -27       -27  Plant C.csv  W29/2021   

       No of OOS weeks  
5                    9  
14                   3  
23                   2  
59                   1  
113                  4  
...                ...  
28049                2  
28067               12  
28076                1  
28211                1  
28265               13  

[278 rows x 21 columns]